### PR TITLE
Transfer ordering

### PR DESCRIPTION
**Pull request by matei-p**
## Reasoning behind the pull request
- 
- 
- 
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/smartContract/processorV2/processV2.go [hunk 1]
```diff
@@ -2369,17 +2369,22 @@ func GetOriginalSenderForTx(tx data.TransactionHandler) []byte {
...
 	outAcc *vmcommon.OutputAccount,
 	tx data.TransactionHandler,
 	txHash []byte,
-) (bool, []data.TransactionHandler) {
+) (bool, []internalIndexedScr) {
 
 	result := sc.createSCRFromStakingSC(outAcc, tx, txHash)
 	if !check.IfNil(result) {
-		return false, []data.TransactionHandler{result}
+		return false, []internalIndexedScr{{result, 0}}
```
Comment (sasurobert) [inline process/smartContract/processorV2/processV2.go:2387]: we need real index - nextIndex here as well.

File: process/smartContract/processorV2/vmOutputAccountsProcessor.go [hunk 2]
```diff
@@ -87,10 +88,28 @@ func (oap *VMOutputAccountsProcessor) Run() (bool, []data.TransactionHandler, er
...
 
+	sortIndexedScrs(indexedSCResults)
+
+	scResults := unwrapSCRs(indexedSCResults)
+
 	return createdAsyncCallback, scResults, nil
 }
 
-func (oap *VMOutputAccountsProcessor) processOutputTransfersStep(outAcc *vmcommon.OutputAccount, createdAsyncCallback bool, scResults []data.TransactionHandler) (bool, []data.TransactionHandler) {
+func unwrapSCRs(indexedSCResults []internalIndexedScr) []data.TransactionHandler {
```
Comment (sasurobert) [inline process/smartContract/processorV2/vmOutputAccountsProcessor.go:98]: nice one :)

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/eei.go [hunk 3]
```diff
@@ -802,3 +805,10 @@ func (host *vmContext) CleanStorageUpdates() {
 func (host *vmContext) IsInterfaceNil() bool {
 	return host == nil
 }
+
+// NextOutputTransferIndex returns next available output transfer index
+func (context *vmContext) NextOutputTransferIndex() uint32 {
```
Comment (sasurobert) [inline vm/systemSmartContracts/eei.go:810]: move this function befor IsInterfaceNil

File: vm/systemSmartContracts/eei.go [hunk 4]
```diff
@@ -802,3 +805,10 @@ func (host *vmContext) CleanStorageUpdates() {
...
+	index := context.crtTransferIndex
+	context.crtTransferIndex++
```
Comment (sasurobert) [inline vm/systemSmartContracts/eei.go:812]: set crtTransferIndex to 0 on CleanCache function.

File: vm/systemSmartContracts/eei.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go:810]: make if func (host *vmContext) to be inline with the whole file.

File: vm/systemSmartContracts/eei.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/eei.go]: This is not done.

**Review by sasurobert — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**